Feedback

bitfield opened the pull request:
:wave:! GitHub Classroom created this pull request as a place for your teacher to leave feedback on your work. It will update automatically. **Don’t close or merge this pull request**, unless you’re instructed to do so by your teacher.

In this pull request, your teacher can leave comments and feedback on your code. Click the **Subscribe** button to be notified if that happens.

Click the **Files changed** or **Commits** tab to see all of the changes pushed to `master` since the assignment started. Your teacher can see this too.

<details>
<summary>
<strong>Notes for teachers</strong>
</summary>

Use this PR to leave feedback. Here are some tips:

- Click the **Files changed** tab to see all of the changes pushed to `master` since the assignment started. To leave comments on specific lines of code, put your cursor over a line of code and click the blue **+** (plus sign). To learn more about comments, read “[Commenting on a pull request](https://docs.github.com/en/github/collaborating-with-issues-and-pull-requests/commenting-on-a-pull-request)”.
- Click the **Commits** tab to see the commits pushed to `master`. Click a commit to see specific changes.
- If you turned on autograding, then click the **Checks** tab to see the results.
- This page is an overview. It shows commits, line comments, and general comments. You can leave a general comment below.

For more information about this pull request, read “[Leaving assignment feedback in GitHub](https://docs.github.com/education/manage-coursework-with-github-classroom/leave-feedback-with-pull-requests)”.
</details>


Subscribed: @dch

bitfield committed:
Setting up GitHub Classroom Feedback

dch committed:
vim-go auto fmt ftw

dch committed:
ex3 subtract tests pass!

dch commented:
## build etc

- neovim + [fatih/vim-go](https://github.com/fatih/vim-go/wiki/Tutorial)
- leader + b|r|t build|run|test
- vim-go I can jump to typedefs somehow? funcs yes, gc| c-] maybe not sure.  c-t or c-o to pop the stack
- `[[|]]` work too, nice, K for doc/type notes. I'd like these to float in neovim

## testing

- how to get vim to take me to https://pkg.go.dev/testing for example? found it in zeal
- how to handle `div/0` which is a panic? tried and failed in dff780c
- https://yourbasic.org/golang/recover-from-panic/ has some notes

dch committed:
divide and conquer

dch committed:
go forth and multiply, badly

dch committed:
i said, multiply, and pass thy tests

dch committed:
add div by zero case

go test -v
=== RUN   TestAdd
=== PAUSE TestAdd
=== RUN   TestSubtract
=== PAUSE TestSubtract
=== RUN   TestMultiply
=== PAUSE TestMultiply
=== RUN   TestDivide
=== PAUSE TestDivide
=== RUN   TestDivideByZero
=== PAUSE TestDivideByZero
=== CONT  TestAdd
=== CONT  TestDivide
--- PASS: TestAdd (0.00s)
--- PASS: TestDivide (0.00s)
=== CONT  TestSubtract
--- PASS: TestSubtract (0.00s)
=== CONT  TestDivideByZero
=== CONT  TestMultiply
=== CONT  TestDivideByZero
    calculator_test.go:49: want 0.000000, got +Inf
--- FAIL: TestDivideByZero (0.00s)
--- PASS: TestMultiply (0.00s)
FAIL
exit status 1
FAIL	calculator	0.005s

dch committed:
try to catch panic in tests

--- PASS: TestSubtract (0.00s)
=== CONT  TestDivideByZero
    calculator_test.go:56: want 0.000000, got +Inf
    calculator_test.go:49: dammit no panic
--- FAIL: TestDivideByZero (0.00s)
FAIL
exit status 1
FAILcalculator0.005s

dch committed:
its not panicking just +Inf is also a "number"

dch committed:
pass errors on divide back to caller

dch committed:
calc: extend testing using a slice of test cases

I expect this changes our parallel testing a wee bit

dch committed:
handle div/zero errors

write the test down in english!!!!
squash the booleans for great success!!

dch committed:
- how do we pass an empty slice in?
- answer is don't pass it at all, go will fill in the blanks
- var... is the unroll operator, used in function heads and to pass in
  parameters where the length of the slice may actually be zero
- also used as concatenation operator

dch committed:
go doesn't have methods on slices so build your own iterator and sum it
up

dch committed:
add a reference to a function our struct (function field)

this is easy as go does the hard work and figures it all out
we now need to invoke the function in our test loop

dch committed:
move testing into a loop, and pass function name and ref in as parameters to the struct

dch committed:
teach subtract to handle variadic parameters

- use unroll operator in function invocations
- add variadic tests

dch committed:
add variadic capability & tests to multiply

dch committed:
remove surplus tests

dch committed:
add square root and tests

dch committed:
combine testDivide cases by using a test struct and a bool flat

dch committed:
try property-based testing

dch committed:
try one way to print out test cases

$ go test -v
...
failed 8.98846567431158e+307, got +Inf, +Inf, +Inf
! : Falsified after 3964 passed tests.
ARG_0: 8.98846567431158e+307
ARG_0_ORIGINAL (1000 shrinks): 1.6254893234507023e+308
Elapsed time: 241.441488ms
    properties.go:57: failed with initial seed: 1609892848820197171
--- FAIL: TestProperties (0.24s)
=== CONT  TestAddSubMul
=== CONT  TestNegativeSquareRoot
--- PASS: TestAddSubMul (0.00s)
--- PASS: TestNegativeSquareRoot (0.00s)
=== CONT  TestDivide
--- PASS: TestDivide (0.00s)
=== CONT  TestSquareRoot
--- PASS: TestSquareRoot (0.00s)
FAIL
exit status 1
FAIL	calculator	0.247s

bitfield commented on calculator.go:
```diff
@@ -12,10 +12,11 @@ func Add(a, b float64, extras ...float64) float64 {
+// initial number, after subtracting the remaining numbers from the
+// first.
+func Subtract(a, b float64, extras ...float64) float64 {
+	return a - Add(0, b, extras...)
```
Neat!

dch committed:
add Parser and tests for future evaluation function

dch committed:
convert Parser to Evaluate for great good

dch committed:
test: tidy up logic a spot

dch committed:
sqrt: test fail

dch committed:
sqrt: test fail

bitfield committed:
refactor error testing logic
